refactor(components): extract Header, Footer, and Main components fro…

MarikaBBB opened the pull request:
…m HomePage

Refactor Homepage Components & Add Search Icon:

- Moved the Header and Footer components into separate .tsx files to improve maintainability and reuse.
- Created Main.tsx to hold the main content of the homepage.
- Updated HomePage.tsx to assemble the homepage by linking Header, Main, and Footer components.
- Installed react-icons package

MarikaBBB committed:
refactor(components): extract Header, Footer, and Main components from HomePage

MarikaBBB committed:
feat(header): add user/shopping icons to Header component

MarikaBBB commented:
I improved the Header interface. 

Added user (FaUserCircle) and shopping bag (FaShoppingBag) icons using the react-icons library.

Styling:
- The logo is on the left, like before.
- User and shopping icons are on the right, with a hover effect.
- A promotional "banner " below the main header bar highlighting "FREE UK DELIVERY ON ALL ORDERS (International delivery available)".

Other changes:
removed` import React from "react";` from the ProductCard.tsx

MarikaBBB committed:
refactor(main): remove button from Main component

MarikaBBB committed:
feat: create NavBar component

MarikaBBB commented:
In this pull request, I created the NavBar component and look into the database schema to understand the main sections/pages that the NavBar should include. Each link in the NavBar now leads to corresponding sections (e.g., Vinyls, CDs, Genres, Artists, Labels, New Releases, etc.), which reflects what we defined in the database. 

NOTE: links in the NavBar are not working. This will be a future task (working on setting up the navigation router to connect each link to the corresponding pages/sections)

Other changes - the Header.tsx
Improved accessibility by adding descriptive labels to the user and cart icons

MarikaBBB committed:
refactor(header): extract IconWithText component for reuse

MarikaBBB commented:
Extracted IconWithText logic into its own file (IconWithText.tsx) for better reuse.
Added IconWithTextProps type in IconWithText.tsx to fix TypeScript issues.

Why This Change?
- Reuse: Makes it easier to reuse the IconWithText component in different parts of the app.
- Cleaner Code: Keeps the Header component simple and focused.

maxitect commented on src/components/layout/Footer.tsx:
```diff
@@ -0,0 +1,15 @@
+    return (
+      <footer role="contentinfo" className="bg-gray-800 text-white py-8 w-full">
+        <div className="text-center">
+          <p>Footer</p>
```
```suggestion
          <p>&copy; 2024 Font Hill Records. All rights reserved.</p>
```

MarikaBBB replied:
Thanks :D

maxitect commented on src/components/layout/Footer.tsx:
Maybe have a look at what Khalos did for their footer, it looked great in their presentation

MarikaBBB replied:
Thank you, Max. I will take this into consideration when I work on the footer, but that will be for another day. I want to finish all the header section today.

maxitect commented on src/pages/Homepage.tsx:
To avoid repeating Header/Footer in every page we could put it directly into the App.tsx so the router only changes the main content like this:
```typescript
function App() {
  return (
    <>
      <Header />
        <Router>
          <Routes>
            <Route path="/" element={<Home />} />
            <Route path="/about" element={<About />} />
            <Route path="/contact" element={<Contact />} />
          </Routes>
        </Router>
      <Footer />
    </>
  )
}
```

MarikaBBB committed:
fix(router): wrap entire app in Router to resolve blank page issue

MarikaBBB committed:
feat: create SearchBar component

MarikaBBB commented:
Following Max's suggestion(which I accidentally delete his comments), I made the changes he asked for:

- Refactored App.tsx: Moved the Header and Footer components directly into App.tsx, as he suggested. I understand that by doing this, the router only changes the main content while keeping the header and footer consistent across all pages.

- Resolved Blank Page Issue: After making these changes, the webpage went blank. To fix this, I wrapped the entire application in the Router to ensure proper routing and visibility of all components (Header, Footer, and page routes).

- Updated Footer Component: Changed the text in Footer.tsx from a placeholder `<p>Footer</p>` to the correct content: `<p>&copy; 2024 Font Hill Records. All rights reserved.</p>`. (Thanks Max)

- SearchBar Component: Completed the creation of the SearchBar component. **NOTE** that the logic for the SearchBar will be implemented in a future task.

MarikaBBB committed:
feat(accessibility): improve accessibility in IconWithText and Logo

MarikaBBB committed:
feat(accessibility): enhance accessibility across multiple components

MarikaBBB committed:
chore(tailwind): update colour palette for improved consistency and contrast

MarikaBBB commented:
The main change is he customisation in  tailwind.config.js file to reflect the new colour palette. The purpose is to have a consistent look throughout the application. I added custom colour configurations for:
- primary, primary-dark, and accent colours.
- background variations (light, default, footer)
- text-primary and text-secondary colours to enhance readability and visual hierarchy.

I refractored the App.tsx file to address the issue where the Footer was not visible. I added flex-shrink-0 to the Footer component, making it a non-shrinkable element so it will not be pushed off the screen by other components.

I adjusted the Main component to prevent it from taking excessive vertical space, which was causing the Footer to be hidden from view.

NOTE: We should decide together the look of our website

maxitect commented on src/App.tsx:
```diff
@@ -1,16 +1,29 @@
-        <Route path="/" element={<Homepage />} />
-        {/* Add other routes here */}
-      </Routes>
+      <div className="flex flex-col min-h-screen">
```
Do we need to wrap everything in a div? Is it necessary to assign styles like this? I think it may be possible to remove this div and if we really do need to assign high level styling perhaps it can live in the root div in index.html?

Otherwise our html page will look like this when rendered:
```html
<div id=root> <-- this is the div inside out index.html file where React inserts all of the web content -->
    <div> <-- this is the div in App.tsx that I am suggesting we don't need -->
        <header>
        <main>
        <footer>
    </div>
</div>

MarikaBBB replied:
Hey Max, thanks for the detailed feedback! You are right, I agreed what what you say.
I’ve removed the unnecessary `<div>` wrapper in `App.tsx` and replaced it with a `<main>` element to maintain semantic structure.

maxitect commented on src/App.tsx:
```diff
@@ -1,16 +1,29 @@
-      </Routes>
+      <div className="flex flex-col min-h-screen">
+        <Header />
+        <div className="flex-grow">
```
should this be `<main>` element rather than div? Remember that divs have no semantic html value so best to use something that has semantic meaning such as `<main>` in this case

maxitect replied:
either that or just remove this div entirely and move whatever styling you need inside Main.tsx

MarikaBBB replied:
I kept the `flex-grow` class on `<main>` to ensure that the content area adjusts as needed while keeping the footer at the bottom. I hope this is good thinking.

maxitect commented on src/App.tsx:
```diff
@@ -1,16 +1,29 @@
+      </div>
     </Router>
   );
 };
```
Are you sure that having Header and Footer inside the Router is the issue causing your blank page?

I know for sure that in another project template I am working on with React this solution does work:
```tsx
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import Header from './components/Header/Header'
import Footer from './components/Footer/Footer'
import Home from './pages/Home'
import About from './pages/About'
import Contact from './pages/Contact'
import './App.css'

function App() {
  return (
    <>
      <Header />
        <Router>
          <Routes>
            <Route path="/" element={<Home />} />
            <Route path="/about" element={<About />} />
            <Route path="/contact" element={<Contact />} />
          </Routes>
        </Router>
      <Footer />
    </>
  )
}

export default App
```

Page renders like this:
![image](https://github.com/user-attachments/assets/d0a6468c-6dc3-48e8-af12-61f10c79fdcf)

MarikaBBB replied:
Thank you once again for the feedback; it was extremely helpful.

maxitect commented on src/components/common/Button.tsx:
```diff
@@ -20,10 +20,16 @@ const Button = ({
   ...props
 }: ButtonProps) => {
+  
+  const baseClasses = 'px-4 py-2 rounded text-white transition duration-300 focus:outline-none focus:ring';
+
+  const colorClasses = disabled
+  ? 'bg-background-light cursor-not-allowed'
+  : 'bg-primary hover:bg-primary-dark focus:bg-primary-dark';
```
This is very neat saving the classes as variables to tidy up the main component area, good idea!

MarikaBBB replied:
Hey Max, thanks for your feedback! I was working on improving accessibility by changing the button's colors when it's focused, hovered, or disabled to give better visual feedback. It was quite a challenging task and it took me a while to do so, especially making sure it's accessible for screen readers and keyboard users. However, I haven't tested it yet, and I need to learn how to do that.

maxitect commented on src/components/layout/Footer.tsx:
```diff
@@ -0,0 +1,18 @@
+      role="contentinfo"
+      className="bg-background-footer text-text-primary py-8 w-full"
+    >
+      <div className="text-center">
+        <p>&copy; 2024 Font Hill Records. All rights reserved.</p>  
+        </div>
```
```suggestion
        <p className="text-center">&copy; 2024 Font Hill Records. All rights reserved.</p>  
```
I think we can avoid using another div here with this solution - or else this styling can be applied to the footer generally. Until we get more content in the footer when a div might be useful lets keep it simple?

MarikaBBB replied:
Hey Max, thanks for the suggestion! I’ve removed the unnecessary <div>

maxitect commented on src/components/layout/NavBar.tsx:
```diff
@@ -0,0 +1,49 @@
+          </Link>
+        </li>
+        <li>
+          <Link to="/shipping" className="text-text-primary hover:text-primary">
```
considering that the exact same styling is being applied to each Link component here perhaps you can use the same method that you used in Button.tsx and save the styling string inside a variable at the top and call it in each instance here. That way if we wanted to change it we only need to do so in the variable at the top to apply to all instances.

MarikaBBB replied:
Hey Max, thanks for pointing that out! You're right, I hadn't thought about using a variable to store the styles. I’ve now refactored the NavBar component to use the linkStyles variable. This makes it much easier to manage and update in the future.

maxitect commented on src/pages/Homepage.tsx:
```diff
@@ -1,78 +1,14 @@
-          <p>Footer</p>
-        </div>
-      </footer>
+    <div className="min-h-screen bg-background-light w-full flex flex-col">
```
Should we simplify the styling somehow? I can see another div another div with styling here I think it means our webpage will have 3 levels of nested div styling, is it required or can it be reduced somehow? Just a though as I think our page will look like this now:
```html
<div id=root>
    <div>
        <header>
        <div> //this is the div in App.tsx
            <div> //this is the div present here
                <main>
            </div>
        </div>
        <footer>
    </div>
</div>
```

MarikaBBB replied:
Hey Max, I changed the <div> to a <main> tag instead. I’ve commented it out for now because I’m deciding whether to remove the wrapper here in HomePage or in App.tsx. I’ll take a closer look at the layout and make sure to simplify it

Thanks for the suggestion — I’ll finalise it soon!

maxitect commented on tailwind.config.js:
Great use of tailwind for our custom colours!

MarikaBBB committed:
fix(App.tsx): removed unnecessary div and used semantic main tag

MarikaBBB committed:
refactor(footer): remove unnecessary div

MarikaBBB committed:
refactor(navbar): extract common link styles into variable for reuse

MarikaBBB committed:
refactor(homepage): replace div with main and comment out pending final decision

MarikaBBB committed:
refactor(homepage): remove unnecessary main wrapper

maxitect commented on src/components/layout/Main.tsx:
```diff
@@ -0,0 +1,63 @@
+      className="items-center justify-center text-center px-4 py-4 flex flex-col"
+      role="main"
+    >
+      <h1 className="text-4xl font-bold mb-4 text-text-primary">
+        Welcome to Font Hill Records
+      </h1>
```
```suggestion
      <h2 className="text-4xl font-bold mb-4 text-text-primary">
        Welcome to Font Hill Records
      </h2>
```
We have an h1 in the header (inside Logo.tsx) and we there can only be 1 h1 on the page

maxitect commented on src/components/layout/Main.tsx:
```diff
@@ -0,0 +1,63 @@
+      </p>
+
+      <section className="mb-12 w-full" aria-labelledby="new-on-store">
+      <h2 id="new-on-store" className="text-2xl font-semibold mb-4 text-text-primary">
+          New on Store</h2>
```
```suggestion
      <h3 id="new-on-store" className="text-2xl font-semibold mb-4 text-text-primary">
          New on Store</h3>
```
If the above is h2, then this should be h3

maxitect commented on src/components/layout/Main.tsx:
```diff
@@ -0,0 +1,63 @@
+  };
+
+  return (
+    <main
```
We now have 2 mains (with the main in App.tsx), so this should probably be a div if we need it - or could we move the styling into App.tsx or vice-versa?

maxitect commented on src/pages/Homepage.tsx:
What is this file for if all it does is import Main.tsx? Perhaps this should be deleted and Main.tsx renamed to Home.tsx if we are using Router we don't need a "Main.tsx"

maxitect reviewed (changes requested):
Looks great, few last comments on accessibility and file management and I think we can merge this, thanks again Marika great work!

MarikaBBB committed:
refactor: changed man to div

MarikaBBB committed:
refactor: changed h1 to h2 to improve accessibility

Co-authored-by: Maxime Downe <81163597+maxitect@users.noreply.github.com>

MarikaBBB committed:
refractor: changed h2 to h3

Co-authored-by: Maxime Downe <81163597+maxitect@users.noreply.github.com>

MarikaBBB committed:
refactor: deleted Hompage.tsx and renamed the file Man to Home

MarikaBBB committed:
refactor: moved Main to page folder

maxitect reviewed (approved):
Perfect, thanks Marika it looks great!
